### ISSUE TITLE
Text to speech should ignore hidden content

### ISSUE THREAD

**Issue by pkra** [edrlab/thorium-reader#1559]
Something like `<p>This is <span hidden="hidden">hidden</span> text</p>` will be voiced "This is hidden text".

**Comment by pkra** [comment]
As mentioned https://github.com/edrlab/thorium-reader/issues/1501#issuecomment-857577821, I realize this leads down the rabbit hole of accname computation. On electron, you might be able to leverage chromium's experimental AOM features (currently behind a flag) but there are also JS implementations.

**Comment by danielweck** [comment]
Thank you for pointing this out, Peter.
Yes, that's quite a "rabbit hole" indeed! ;)
(e.g. CSS `display: none`, `visibility: hidden`, `opacity: 0`, `transform: translateX(-9999px)`, etc.)

However the `hidden` attribute expresses a very clear intent, so I think can be translated into a rather "safe" processing rule (i.e. skip the HTML fragment during the DOM walk which builds the sequence of utterances).

**Comment by pkra** [comment]
> CSS display: none, visibility: hidden, opacity: 0, transform: translateX(-9999px)

Of these, only `display:none` would remove the element from the accessibility tree -- though technically aria-labelledby / describedby can point to `display:none`'d content, too. 

> However the hidden attribute expresses a very clear intent, so I think can be translated into a rather "safe" processing rule

I don't think it's quite as clear cut. CSS can override `hidden`, e.g., `<span hidden="hidden" style="display:block">not hidden</span>`, `.override[hidden] { display: block}`, `[hidden]:focus {display: block}`.

While I found this in a situation where the `hidden` content can always be ignored, I immediately thought of the [old checkbox sibling hack](https://css-tricks.com/the-checkbox-hack/) and other use cases for CSS-only interaction and adjustment. They would require a more refined read-aloud technique. 

Anyway, maybe the read-aloud feature doesn't need to support these scenarios.

**Comment by SimonPRH** [comment]
Hello. This is an issue for us too.

We’re currently testing our books with different screen reading software (Apple Voiceover and Thorium’s built-in TTS so far, next Jaws and NVDA), and noticed something unexpected.

We use a single NAV file as both the sidebar, and internal table of contents for our reflowable ebooks. We use a class called "hidden_content" with display: none to hide some TOC entries from the internal TOC, and we use that class along with hidden=”hidden” to hide the pagelist and landmarks sections. While Apple Voiceover on Apple Books respects this, Thorium’s TTS reads all of these hidden entries regardless.

We would use "aria-hidden=true'" to explicitly exclude screen reader clients, but W3 guidance suggests we shouldn't use it on 'focusable elements'. An anchor on its own is not focusable (<a>...</a>), but becomes focusable when you add an href attribute to it (<a href="...">...</a>). So we’re not sure if it’s permissible to do this.

**Comment by danielweck** [comment]
Fixed via https://github.com/edrlab/thorium-reader/commit/c250f59cb5d2116f427937ffa6dd18a5e23242b1

Build in progress: https://github.com/edrlab/thorium-reader/actions/runs/1594821427

(will be available here in a few minutes: https://github.com/edrlab/thorium-reader/releases )
